### PR TITLE
feat!(...): update parser + schema

### PR DESCRIPTION
**Pull request by yosefAlsuhaibani**
PR checklist:

- [x] Purpose of the code is [evident to future readers](https://semgrep.dev/docs/contributing/contributing-code/#explaining-code)
- [ ] Tests included or PR comment includes a reproducible test plan
- [ ] Documentation is up-to-date
- [ ] A changelog entry was [added to changelog.d](https://semgrep.dev/docs/contributing/contributing-code/#adding-a-changelog-entry) for any user-facing change
- [ ] Change has no security implications (otherwise, ping security team)

If you're unsure about any of this, please see:

- [Contribution guidelines](https://semgrep.dev/docs/contributing/contributing-code)!
- [One of the more specific guides located here](https://semgrep.dev/docs/contributing/contributing/)

### CODE REVIEW

File: libs/networking/Http.ml
Comment (aryx) [inline libs/networking/Http.ml]: This will be in conflict with @ajbt200128 who is also modifying this file to use another Http library ...
We need to figure things out here and use just one if possible

Comment (aryx) [reply libs/networking/Http.ml]: But yes we can move to libs/networking/

Comment (aryx) [reply libs/networking/Http.ml]: We can also have Http_helpers1.ml and Http_helpers2.ml for now, where one use cohttp and ther other piaf
(or maybe http_helpers_piaf.ml and http_helpers_cohttp.ml)

File: dune-project
Comment (aryx) [inline dune-project]: maybe networking if it's in libs/networking

File: src/core/Rule.ml
Comment (aryx) [inline src/core/Rule.ml]: What do you need to figure out?
Maybe be clearer in the comment on what is the TODO

File: src/core/Rule.ml
Comment (aryx) [inline src/core/Rule.ml]: better define intermediate structures for those, like request_spec, response_spec (or just request and response)

File: src/core/Rule.ml
Comment (aryx) [inline src/core/Rule.ml]: Can we use Uri.t here?

File: src/core/Rule.ml
Comment (aryx) [inline src/core/Rule.ml]: Can we use a more precise type?
The http libraries usually define this like Get | Post | ...

File: src/core/Rule.ml
Comment (aryx) [inline src/core/Rule.ml]: same, better to have intermediate record for response since I guess it will grow

File: src/engine/Match_rules.ml
Comment (aryx) [inline src/engine/Match_rules.ml]: so I guess this will be handled at some point in semgrep-pro?

File: src/osemgrep/cli_ci/Github_metadata.ml
Comment (aryx) [inline src/osemgrep/cli_ci/Github_metadata.ml]: No!
Fix it.
The code was working before, just define your own http_helpers.ml functions to not bother
existing code.

File: libs/networking/Http.ml
Comment (aryx) [reply libs/networking/Http.ml]: Also please keep as Http_helpers.ml, so it's clearer it's just some wrappers over another
official http library, otherwise it looks like we use an official Http robust library.

File: src/osemgrep/cli_login/Login_subcommand.ml
Comment (aryx) [inline src/osemgrep/cli_login/Login_subcommand.ml]: Might be better to restore Http_helpers_piaf.ml or something.

File: src/parsing/Parse_rule.ml
Comment (aryx) [inline src/parsing/Parse_rule.ml:1714]: please use a better name, maybe semgrep_internal_secrets.
Otherwise I have no idea wiaht postprocessor mean/

**Review by aryx — CHANGES_REQUESTED** [review]

File: src/parsing/Parse_rule.ml
Comment (emjin) [reply src/parsing/Parse_rule.ml:1714]: This was a compromise with Lewis to make him ok with merging the semgrep changes into the public facing repo; it's intentionally vague.

File: src/engine/Match_rules.ml
Comment (yosefAlsuhaibani) [reply src/engine/Match_rules.ml]: Yes; evaluation happens there.

File: src/core/Rule.ml
Comment (yosefAlsuhaibani) [reply src/core/Rule.ml]: I think thats an old TODO; when I was drafting syntaxes; will delete that line

File: src/osemgrep/cli_ci/Github_metadata.ml
Comment (yosefAlsuhaibani) [reply src/osemgrep/cli_ci/Github_metadata.ml]: Oh my bad 😭 I left this as a placeholder when I was moving between libraries, will fix...

File: src/core/Rule.ml
Comment (yosefAlsuhaibani) [reply src/core/Rule.ml]: Added `request_spec` && `response_spec`!

File: src/core/Rule.ml
Comment (yosefAlsuhaibani) [reply src/core/Rule.ml]: Changed it to `meth : [`GET | `POST]`.

File: src/osemgrep/cli_login/Login_subcommand.ml
Comment (yosefAlsuhaibani) [reply src/osemgrep/cli_login/Login_subcommand.ml]: I'll just restore it to `Http_helpers`

File: src/parsing/Parse_rule.ml
Comment (emjin) [reply src/parsing/Parse_rule.ml:1714]: @yosefAlsuhaibani We should add a TODO to make this more specific

File: src/core/Rule.ml
Comment (emjin) [inline src/core/Rule.ml]: What are each of these strings? It would be nice to have comments or make new types

File: src/core/Rule.ml
Comment (emjin) [inline src/core/Rule.ml]: Needs more explanation. What does postprocessor-patterns refer to? Why is it associated with `secrets:`?

File: src/core/Rule.ml
Comment (emjin) [inline src/core/Rule.ml]: When did this get renamed to `step_mode` and `step_rule` (no `s`)?

File: src/core/Rule.ml
Comment (aryx) [inline src/core/Rule.ml]: typo, spec no sepc.
Also I would remove those _spec prefix (I have a PR where I removed it from extract; I kept the
taint_spec because there's code depending on this name type, and I was lazy, but I think it would
be simpler to remove those _spec suffix from the start here.

File: libs/networking/dune
Comment (aryx) [inline libs/networking/dune]: better networking for the name, to match the directory
(and so that later we can add helpers for other networking stuff, not just http)

File: src/parsing/Parse_rule.ml
Comment (aryx) [reply src/parsing/Parse_rule.ml:1714]: But we have lots of use of `Secrets in the code ... in the public repo ... even this discussion is in the public repo
so I don't really get obfuscating it in the rule.

**Review by aryx — DISMISSED** [review]

File: src/parsing/Parse_rule.ml
Comment (emjin) [inline src/parsing/Parse_rule.ml:308]: Minor: I would just call this "method"

File: src/parsing/Parse_rule.ml
Comment (emjin) [reply src/parsing/Parse_rule.ml:1714]: The rules are user facing, so I believe the goal is to keep users from getting used to `mode: secrets` (since this is experimental and we think it will change) and to keep it a little quieter, since people are more likely to see rules than look at our code base. I don't think we're trying that hard to keep it secret though, all our paying customers know it's coming soon and it's not like they're under NDA.

Lewis would prefer we also don't use secrets in the code but he's ok with it if it would be very annoying for us as maintainers.

File: libs/networking/Http.ml
Comment (yosefAlsuhaibani) [reply libs/networking/Http.ml]: We're switching to only Cohttp + moving networking to a libs + changing `Http.*` to `Http_helpers.*`

File: src/core/Rule.ml
Comment (yosefAlsuhaibani) [reply src/core/Rule.ml]: I think this was due to a kerfuffle a merge between my branch & develop; will change them back

Comment (yosefAlsuhaibani) [reply src/core/Rule.ml]: I think this is due to a kerfuffle when I was merging into develop; reverted back

File: src/core/Rule.ml
Comment (yosefAlsuhaibani) [reply src/core/Rule.ml]: Added bits of documentation; should I expand a bit more?

File: src/core/Rule.ml
Comment (yosefAlsuhaibani) [reply src/core/Rule.ml]: Removed _spec

File: src/core/Rule.ml
Comment (yosefAlsuhaibani) [reply src/core/Rule.ml]: I think I've updated the types since then to add more context to what this type encodes; `headers` field might be a bit ambigous since its still just `string * string`, but I could update it to a `headers : { name : string; value : string } list`; Or `Cohttp.Headers.t`.

File: src/core/Rule.ml
Comment (yosefAlsuhaibani) [reply src/core/Rule.ml]: Add intermediate records for `response` & `request`

File: src/core/Rule.ml
Comment (emjin) [reply src/core/Rule.ml]: A record would be nice!

File: src/core/Rule.ml
Comment (aryx) [inline src/core/Rule.ml:282]: Great comment!
Also might be useful to explain why we don't reuse the request type of http libraries like cohttp, which have their own request type.

File: src/core/Rule.ml
Comment (aryx) [inline src/core/Rule.ml]: you can remove the comment, it's not useful anymore

File: src/core/Rule.ml
Comment (aryx) [inline src/core/Rule.ml]: same, remove old comment.

**Review by aryx — COMMENTED** [review]
LGTM.
I'll let @emjin approve.

**Comment by aryx** [comment]
Lots of CI failures to fix though.

**Review by emjin — APPROVED** [review]
Let's merge it!

**Pull request merged**
